feat: add Confirmation Signature Redesign ui_customization event prop

digiwand opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Adds ui_customization: 'redesigned_confirmation' to the Signature <Approved|Rejected|Canceled> events

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24182?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/23555

## **Manual testing steps**

1. set `ENABLE_CONFIRMATION_REDESIGN=true` in .metamaskrc
2. enable Experimental > Redesign Confirmation in settings
3. enable Participate in MetaMetrics in settings
4. test signatures (excluding Sign-in with Ethereum)
5. observe Signature <Rejected|Canceled|Approved> event with new ui_customization


## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

digiwand committed:
chore: add newlines in useCurrentConfirmation

digiwand committed:
refactor: add const ConfirmationRedesignTypes

digiwand committed:
feat: add confirmation redesign ui_customizations to metric event

digiwand committed:
test: add createRPCMethod confirm redesign ui_customization event prop tests

digiwand committed:
Merge branch 'develop' into feat-add-conf-redesign-sign-ui-customizations-metametrics

pedronfigueiredo commented on app/scripts/lib/createRPCMethodTrackingMiddleware.test.js:
```diff
@@ -416,6 +429,77 @@ describe('createRPCMethodTrackingMiddleware', () => {
+      expect(trackEvent).toHaveBeenCalledTimes(2);
+
+      expect(trackEvent.mock.calls[1][0]).toMatchObject({
+        category: 'inpage_provider',
```
```suggestion
        category: MetaMetricsEventCategory.InpageProvider,
```

digiwand replied:
thanks! 👍🏼  was using the same pattern in the test. updated the 10 instances in the file [223a699](https://github.com/MetaMask/metamask-extension/pull/24182/commits/223a6997243205354f41891e18dd936ba33d5226)

pedronfigueiredo commented on app/scripts/lib/createRPCMethodTrackingMiddleware.test.js:
```diff
@@ -416,6 +429,77 @@ describe('createRPCMethodTrackingMiddleware', () => {
+      expect(trackEvent).toHaveBeenCalledTimes(2);
+
+      expect(trackEvent.mock.calls[1][0]).toMatchObject({
+        category: 'inpage_provider',
```
same as above

digiwand replied:
thanks! 👍🏼  was using the same pattern in the test. updated the 10 instances in the file [223a699](https://github.com/MetaMask/metamask-extension/pull/24182/commits/223a6997243205354f41891e18dd936ba33d5226)

digiwand committed:
refactor: 'inpage_provider' -> MetaMetricsEventCategory.InpageProvider

jpuri reviewed (dismissed):
🚀

digiwand commented on ui/pages/confirmations/hooks/useCurrentConfirmation.ts:
```diff
@@ -48,27 +51,28 @@ const useCurrentConfirmation = () => {
-        pendingConfirmation.type !== ApprovalType.EthSignTypedData
-      ) {
+
+      if (!ConfirmationRedesignTypes.includes(pendingConfirmation.type)) {
```
this causes a TS error 
<img width="1045" alt="Screenshot 2024-04-24 at 3 37 16 PM" src="https://github.com/MetaMask/metamask-extension/assets/20778143/65aa3c0b-ecc7-4de2-b570-781ddf8e93cd">

I'm working on updating the expected ApprovalRequest type through our core packages
https://github.com/MetaMask/core/pull/4209

digiwand replied:
Updated the code to allow both ApprovalType and string types in the meantime [096a49a](https://github.com/MetaMask/metamask-extension/pull/24182/commits/096a49a3887f7c8f650bd15ea7db4307ef0782c5)

digiwand committed:
fix: TS lint error for ConfirmationRedesignTypes type
ApprovalController > ApprovalRequest type is currently set as string rather than ApprovalType

mcmire commented on ui/pages/confirmations/utils/confirm.ts:
```diff
@@ -2,6 +2,11 @@ import { ApprovalRequest } from '@metamask/approval-controller';
 import { ApprovalType } from '@metamask/controller-utils';
 import { Json } from '@metamask/utils';
 
+export const ConfirmationRedesignTypes: (ApprovalType | string)[] = [
+  ApprovalType.PersonalSign,
+  ApprovalType.EthSignTypedData,
+];
```
From a type perspective, we might want to keep this the same as it is now. In fact, when it comes to arrays like this I'm curious whether this ought to be defined using `as const` rather than using a type annotation here, so that TypeScript has more information on what the allowed types are when this array is used (for instance, if we were to loop over this list we might want to know that the type of each element is `ApprovalType.PersonalSign | ApprovalType.EthSignTypedData` rather than just `string`). So I would suggest:

``` suggestion
export const ConfirmationRedesignTypes = [
  ApprovalType.PersonalSign,
  ApprovalType.EthSignTypedData,
] as const;

mcmire commented on ui/pages/confirmations/hooks/useCurrentConfirmation.ts:
```diff
@@ -48,27 +51,28 @@ const useCurrentConfirmation = () => {
-        pendingConfirmation.type !== ApprovalType.EthSignTypedData
-      ) {
+
+      if (!ConfirmationRedesignTypes.includes(pendingConfirmation.type)) {
```
Instead of widening the type for `ConfirmationRedesignTypes` overall, what do you think about assigning `ConfirmationRedesignTypes` to a local variable here so that we can temporarily widen the type? Something like

``` typescript
const confirmationRedesignTypes: string[] = ConfirmationRedesignTypes;

if (!confirmationRedesignTypes.includes(pendingConfirmation.type)) {
```

(feel free to move the variable higher up)

MajorLift replied:
There's no type alteration necessary if we simply avoid `includes`. 

```suggestion
      if (!ConfirmationRedesignTypes.find((type) => type === pendingConfirmation.type)) {
```

The underlying cause of the error in this PR that `includes` doesn't handle things well when handed string literal types and string types together.

I agree with @mcmire on everything else, especially not widening `ConfirmationRedesignTypes`.

digiwand replied:
Cool solutions and thanks for sharing @mcmire & @MajorLift! Makes sense and looks good to me

digiwand committed:
refactor: update ConfirmationRedesignTypes TS

Co-authored-by: Elliot Winkler <elliot.winkler@gmail.com>

digiwand committed:
refactor: ConfirmationRedesignTypes use find to match type values

Co-authored-by: Jongsun Suh <jongsun.suh@icloud.com>

digiwand committed:
fix: useCurrentConfirmation lint error newline

digiwand committed:
fix: TS Lint useCurrentConfirmation pendingConfirmation.type

matthewwalsh0 commented on ui/pages/confirmations/utils/confirm.ts:
```diff
@@ -2,6 +2,11 @@ import { ApprovalRequest } from '@metamask/approval-controller';
 import { ApprovalType } from '@metamask/controller-utils';
 import { Json } from '@metamask/utils';
 
+export const ConfirmationRedesignTypes = [
```
Minor, but since this is an exported constant rather than a type or enum, should we go all capitals and underscores?

digiwand replied:
I've been confused about this myself. I've written these as all capitals and underscores in the past though noticed our repo has other patterns. 

I followed the casing from const below.

I also haven't been able to find a reference for const naming in our contributor docs 
https://github.com/MetaMask/contributor-docs/blob/main/docs/javascript.md

- 

Happy to update these to capitals and underscores since it is a common convention and this is an immutable const
ref:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const
> common convention is to use all-uppercase letters, especially for primitives because they are truly immutable.

digiwand replied:
updated https://github.com/MetaMask/metamask-extension/pull/24182/commits/9f10833cb46939a0ec20fdd3b813a5a95cced530 &  https://github.com/MetaMask/metamask-extension/pull/24182/commits/80475d93ab25718a9d83db8bafefc4793a45e32d

matthewwalsh0 commented on app/scripts/lib/createRPCMethodTrackingMiddleware.js:
```diff
@@ -252,6 +255,18 @@ export default function createRPCMethodTrackingMiddleware({
+          ConfirmationRedesignTypes.includes(method);
+
+        if (isRedesignedConfirmation) {
+          eventProperties.ui_customizations = (
```
Minor, but would it condense the syntax a little to use a spread operator such as:

```
eventProperties.ui_customizations = [
    ...eventProperties.ui_customizations,
    MetaMetricsEventUiCustomization.RedesignedConfirmation
];
```

digiwand replied:
Thanks for bringing this up. The reason I haven't done this is I'm unsure how our data platform implementation handles empty arrays and we wouldn't want to add empty arrays to them. I can look into this more

digiwand replied:
sgtm! updated to
```
eventProperties.ui_customizations = [
  ...(eventProperties.ui_customizations || []),
  MetaMetricsEventUiCustomization.RedesignedConfirmation,
];
```

digiwand committed:
Merge branch 'develop' into feat-add-conf-redesign-sign-ui-customizations-metametrics

digiwand committed:
refactor: pass isConfirmationRedesignEnabled to createRPCMethodTrackingMiddleware instead of preferencesController

digiwand committed:
refactor: confirm redesign check RedesignTransactionTypes

digiwand committed:
refactor: redesign types -> use uppercase underscore naming

digiwand committed:
refactor: SignatureApprovalTypes -> SIGNATURE_APPROVAL_TYPES

digiwand committed:
test:fix: createRPCMethodTrackingMiddleware w/ isConfirmationRedesignEnabled prop

digiwand commented on app/scripts/metamask-controller.js:
```diff
@@ -4931,6 +4931,7 @@ export default class MetamaskController extends EventEmitter {
         ),
         getAccountType: this.getAccountType.bind(this),
         getDeviceModel: this.getDeviceModel.bind(this),
+        preferencesController: this.preferencesController,
```
sounds good to me! updated https://github.com/MetaMask/metamask-extension/pull/24182/commits/000a2ee5a3898e56ffa5c40f86afd620cb362154

digiwand commented on app/scripts/lib/createRPCMethodTrackingMiddleware.js:
```diff
@@ -252,6 +255,18 @@ export default function createRPCMethodTrackingMiddleware({
         ///: END:ONLY_INCLUDE_IF
 
+        const isRedesignedConfirmation =
+          process.env.ENABLE_CONFIRMATION_REDESIGN &&
```
I'm a fan of this! thank you. updated https://github.com/MetaMask/metamask-extension/pull/24182/commits/000a2ee5a3898e56ffa5c40f86afd620cb362154
[2407037](https://github.com/MetaMask/metamask-extension/pull/24182/commits/24070377c89092bca842840f12a82fe248299dde)

digiwand committed:
fix: metamask-controller lint newlines

digiwand committed:
fix: useCurrentConfirmation isConfirmationRedesignType check

digiwand committed:
fix: lint pendingConfirmation?.type

digiwand committed:
refactor: createRPCMethodTrackingMiddleware use spread operator for ui_customizations

digiwand committed:
Merge branch 'develop' into feat-add-conf-redesign-sign-ui-customizations-metametrics

jpuri commented on ui/pages/confirmations/utils/confirm.ts:
```diff
@@ -1,13 +1,23 @@
+export const REDESIGN_APPROVAL_TYPES = [
+  ApprovalType.PersonalSign,
+  ApprovalType.EthSignTypedData,
+] as const;
```
May be also add `Transaction` here

digiwand replied:
hmm true since we just created a ticket to separate the Signature Redesign logic from the ENV variable https://github.com/MetaMask/metamask-extension/issues/24312

I think, to be safe, we can add this following the above ticket

digiwand replied:
sgtm! updated [e904ef2](https://github.com/MetaMask/metamask-extension/pull/24182/commits/e904ef2893fec4a59db599cf4fd7caa862f42672)

jpuri commented on ui/pages/confirmations/hooks/useCurrentConfirmation.ts:
```diff
@@ -47,29 +49,37 @@ const useCurrentConfirmation = () => {
+        ) ||
+        REDESIGN_TRANSACTION_TYPES.find(
+          (type) => type === unconfirmedTransaction?.type,
+        );
```
May be extract above to a util method, it used here also: https://github.com/MetaMask/metamask-extension/pull/24182/files#diff-e78ba84d4b27f8a33233149f7c076f773341ab940a409651508f75a2af6e67eeR263

digiwand replied:
Thanks @jpuri. I did consider this. I'm okay repeating code 2-3 before DRYing. We could create a util method if we feel strong about this 

Q: how would you structure the util method with the two different types to check? particularly, would the idea be to allow 2 different params, one for each type?

jpuri commented on app/scripts/metamask-controller.js:
```diff
@@ -4940,6 +4945,7 @@ export default class MetamaskController extends EventEmitter {
         getAccountType: this.getAccountType.bind(this),
         getDeviceModel: this.getDeviceModel.bind(this),
+        isConfirmationRedesignEnabled,
         snapAndHardwareMessenger: this.controllerMessenger.getRestricted({
```
It may be better to inject preference controller state rather than injecting single property.

digiwand replied:
I did initially pass the preferencesController and then we had a conversation to pass the single prop here
https://github.com/MetaMask/metamask-extension/pull/24182#discussion_r1580908661. Open to hearing more opinions on this

digiwand replied:
I initially passed the preferencesController and then we had a conversation to pass the single prop here
https://github.com/MetaMask/metamask-extension/pull/24182#discussion_r1580908661. Open to hearing more opinions on this

matthewwalsh0 replied:
My thinking was it avoids coupling another entire controller to a specific middleware, plus it allowed us to encapsulate the knowledge of the env in this file, rather than the middleware, and potentially other places in future.

jpuri reviewed (commented):
Changes look great, I left some small feedbacks.

digiwand committed:
feat: add ApprovalType.Transaction to REDESIGN_APPROVAL_TYPES

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -4930,6 +4930,11 @@ export default class MetamaskController extends EventEmitter {
 
+    const isConfirmationRedesignEnabled =
+      process.env.ENABLE_CONFIRMATION_REDESIGN &&
+      this.preferencesController.store.getState().preferences
```
As we're checking the preference each time the middleware is used, should this be a function rather than a static property?

digiwand replied:
great catch! Updated [56a0910](https://github.com/MetaMask/metamask-extension/pull/24182/commits/56a091084771a262da76e74a58e18d53569a7aa8)

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -4930,6 +4930,11 @@ export default class MetamaskController extends EventEmitter {
+    const isConfirmationRedesignEnabled =
+      process.env.ENABLE_CONFIRMATION_REDESIGN &&
+      this.preferencesController.store.getState().preferences
+        .redesignedConfirmations;
```
Just a warning that we have another PR to rename this I believe.

matthewwalsh0 commented on app/scripts/lib/createRPCMethodTrackingMiddleware.js:
```diff
@@ -252,6 +258,18 @@ export default function createRPCMethodTrackingMiddleware({
 
+        const isConfirmationRedesign =
+          isConfirmationRedesignEnabled &&
+          (REDESIGN_APPROVAL_TYPES.find((type) => type === method) ||
```
Should we simplify this to:

`[...REDESIGN_APPROVAL_TYPES, ...REDESIGN_TRANSACTION_TYPES].find(type => type === method)`

digiwand replied:
absolutely! ty [3531395](https://github.com/MetaMask/metamask-extension/pull/24182/commits/3531395431a7824894ca06ce3b5e094bb38bc979)

digiwand committed:
refactor: isConfirmationRedesign splay array

digiwand committed:
Revert "refactor: pass isConfirmationRedesignEnabled to createRPCMethodTrackingMiddleware instead of preferencesController"

This reverts commit 000a2ee5a3898e56ffa5c40f86afd620cb362154.

digiwand committed:
refactor: use isConfirmationRedesignEnabled() in RPC middlware

digiwand committed:
fix:test: stub isConfirmationRedesignEnabled

digiwand committed:
refactor: rn isConfirmationRedesignEnabled -> isConfirmationRedesignEnabledFn

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -4940,6 +4948,7 @@ export default class MetamaskController extends EventEmitter {
         ),
         getAccountType: this.getAccountType.bind(this),
         getDeviceModel: this.getDeviceModel.bind(this),
+        isConfirmationRedesignEnabledFn,
```
Very minor, but we have plenty of `isX` callbacks so the `Fn` might not be needed.

digiwand replied:
👍🏼  works with me [cde5b5e](https://github.com/MetaMask/metamask-extension/pull/24182/commits/cde5b5ee0ffcc9cdc1ae6a1bc155a1f41eec56cb)

digiwand committed:
refactor: rn isConfirmationRedesignEnabledFn -> isConfirmationRedesignEnabled
